python/lsst/obs/pfs/isrTask.py

```diff
@@ -0,0 +1,235 @@
+    """
+    doBrokenRedShutter = pexConfig.Field(dtype=bool, default=False,
+                                         doc="Attempt to correct for a broken red shutter?")
+
+    brokenRedShutter = pexConfig.ConfigField(
+        dtype=BrokenShutterConfig, doc="Broken shutter related configuration options."
+    )
+
+    def validate(self):
+        super().validate()
+
+
+class PfsIsrTask(ipIsr.IsrTask):
+    """Apply common instrument signature correction algorithms to a raw frame.
+
+    We run the vanilla ISR, with the exception of optionally correcting for
+    the effects of a broken red-arm shutter on a PFS spectrograph
+    methods have been split into subtasks that can be redirected
+    appropriately.
+
+    Parameters
+    ----------
+    args : `list`
+        Positional arguments passed to the Task constructor. None used at this time.
+    kwargs : `dict`, optional
+        Keyword arguments passed on to the Task constructor. None used at this time.
+    """
+    ConfigClass = PfsIsrTaskConfig
+    _DefaultName = "isr"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    @pipeBase.timeMethod
+    def runDataRef(self, sensorRef):
+        """Perform instrument signature removal on a ButlerDataRef of a Sensor.
+
+        This method contains the `CmdLineTask` interface to the ISR
+        processing.  All IO is handled here, freeing the `run()` method
+        to manage only pixel-level calculations.  The steps performed
+        are:
+        - Read in necessary detrending/isr/calibration data.
+        - Process raw exposure in `run()`.
+        - Persist the ISR-corrected exposure as "postISRCCD" if
+          config.doWrite=True.
+
+        Parameters
+        ----------
+        sensorRef : `daf.persistence.butlerSubset.ButlerDataRef`
+            DataRef of the detector data to be processed
+
+        Returns
+        -------
+        result : `lsst.pipe.base.Struct`
+            Result struct with component:
+            - ``exposure`` : `afw.image.Exposure`
+                The fully ISR corrected exposure.
+
+        Raises
+        ------
+        RuntimeError
+            Raised if a configuration option is set to True, but the
+            required calibration data does not exist.
+
+        """
+        self.log.info("Performing ISR on sensor %s" % (sensorRef.dataId))
+
+        ccdExposure = sensorRef.get(self.config.datasetType)
+
+        camera = sensorRef.get("camera")
+        if camera is None and self.config.doAddDistortionModel:
+            raise RuntimeError("config.doAddDistortionModel is True "
+                               "but could not get a camera from the butler")
+        isrData = self.readIsrData(sensorRef, ccdExposure)
+
+        result = self.run(ccdExposure, camera=camera, **isrData.getDict())
+
+        if self.config.doBrokenRedShutter and sensorRef.dataId["arm"] == 'r':
+            if self.config.doBrokenRedShutter and self.config.brokenRedShutter.useAnalytic:
```

Comment:
Would be good to check and raise an error if ``doBrokenRedShutter`` is ``True`` and the arm is not 'r'

Reply:
Actually the `+ 10000` is part of this debug code, so I need to remove that too (and the format is wrong -- clearly
I added this check-the-exception code but got distracted before doing the checking).  I'm glad you spotted this.

Reply:
I don't think so, as that's a config for the processing and we're generally processing both b and r